File: LeetCode_Java/designclass/FileSystem.java
package designclass;

import java.util.*;

public class FileSystem {

    private class FileSystemNode {
        String content;
        boolean isFile;
        Map<String, FileSystemNode> children;

        public FileSystemNode() {
            content = "";
            isFile = false;
            children = new HashMap<>();
        }
    }

    FileSystemNode root;

    public FileSystem() {
        root = new FileSystemNode();
    }

    public List<String> ls(String path) {
        FileSystemNode node = traverse(path);
        List<String> result = new ArrayList<>();
        if (node.isFile) {
            int index = path.lastIndexOf('/');
            result.add(path.substring(index + 1));
        } else {
            result.addAll(node.children.keySet());
            Collections.sort(result);
        }
        return result;
    }

    public void mkdir(String path) {
        traverse(path);
    }

    public void addContentToFile(String filePath, String content) {
        FileSystemNode node = traverse(filePath);
        node.isFile = true;
        node.content += content;
    }

    public String readContentFromFile(String filePath) {
        FileSystemNode node = traverse(filePath);
        return node.content;
    }

    private FileSystemNode traverse(String path) {
        FileSystemNode cur = root;
        if (path.equals("/")) {
            return cur;
        }

        String[] paths = path.split("/");
        for (String p : paths) {
            if (p.isEmpty()) continue;
            cur.children.putIfAbsent(p, new FileSystemNode());
            cur = cur.children.get(p);
        }

        return cur;
    }

    public static void main(String[] args) {
        FileSystem fileSystem = new FileSystem();
        System.out.println(fileSystem.ls("/")); // []

        fileSystem.mkdir("/a/b/c");
        fileSystem.addContentToFile("/a/b/c/d", "hello");
        System.out.println(fileSystem.ls("/")); // ["a"]
        System.out.println(fileSystem.readContentFromFile("/a/b/c/d")); // "hello"
    }
}

File: LeetCode_Python/designclass/FileSystem.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_file = False
        self.content = ""


class FileSystem:

    def __init__(self):
        self.root = TrieNode()

    def ls(self, path: str) -> list[str]:
        if path == "/":
            return sorted(self.root.children.keys())

        node = self.find(path)

        if node.is_file:
            return [path.split("/")[-1]]
        else:
            return sorted(node.children.keys())

    def mkdir(self, path: str) -> None:
        self.find(path)

    def addContentToFile(self, filePath: str, content: str) -> None:
        node = self.find(filePath, True)
        file_name = filePath.split("/")[-1]
        if file_name not in node.children:
            node.children[file_name] = TrieNode()
            node.children[file_name].is_file = True
        node.children[file_name].content += content

    def readContentFromFile(self, filePath: str) -> str:
        node = self.find(filePath)
        return node.content

    def find(self, path: str, file: bool = False) -> TrieNode:
        if file:
            parts = path.split("/")[1:-1]
        else:
            parts = path.split("/")[1:]
        node = self.root
        for part in parts:
            if part not in node.children:
                node.children[part] = TrieNode()
            node = node.children[part]
        return node


if __name__ == "__main__":
    solution = FileSystem()

    print("Expected Output : []")
    print("Actual Output : ", solution.ls("/"))
    solution.mkdir("/a/b/c")
    solution.addContentToFile("/a/b/c/d", "hello")
    print("Expected Output : ['a']")
    print("Actual Output : ", solution.ls("/"))
    print("Expected Output : ['b']")
    print("Actual Output : ", solution.ls("/a"))
    print("Expected Output : hello")
    print("Actual Output : ", solution.readContentFromFile("/a/b/c/d"))
    solution.addContentToFile("/a/b/c/d", "world")
    print("Expected Output : 'helloworld'")
    print("Actual Output : ", solution.readContentFromFile("/a/b/c/d"))
